devito/petsc/iet/routines.py

```diff
@@ -439,90 +762,241 @@ def runsolve(self, solver_objs, objs, rhs_callback, injectsolve):
+    def replace_array(self, solver_objs):
+        """
+        VecReplaceArray() is a PETSc function that allows replacing the array
+        of a `Vec` with a user provided array.
+        https://petsc.org/release/manualpages/Vec/VecReplaceArray/
 
-    vec_get_size = petsc_call(
-        'VecGetSize', [solver_objs['x_local'], Byref(solver_objs['localsize'])]
-    )
+        This function is used to replace the array of the PETSc solution `Vec`
+        with the array from the `Function` object representing the target.
 
-    field_from_ptr = FieldFromPointer(
-        target.function._C_field_data, target.function._C_symbol
-    )
+        Examples
+        --------
+        >>> self.target
+        f1(x, y)
+        >>> call = replace_array(solver_objs)
+        >>> print(call)
+        PetscCall(VecReplaceArray(x_local_0,f1_vec->data));
+        """
+        field_from_ptr = FieldFromPointer(
+            self.target.function._C_field_data, self.target.function._C_symbol
+        )
+        vec_replace_array = (petsc_call(
+            'VecReplaceArray', [solver_objs['x_local'], field_from_ptr]
+        ),)
+        return vec_replace_array
 
-    expr = DummyExpr(
-        start_ptr, cast_mapper[(target.dtype, '*')](field_from_ptr) +
-        Mul(target_time, solver_objs['localsize']), init=True
-    )
+    def assign_time_iters(self, struct):
+        return []
 
-    vec_replace_array = petsc_call('VecReplaceArray', [solver_objs['x_local'], start_ptr])
-    return (vec_get_size, expr, vec_replace_array)
 
+class TimeDependent(NonTimeDependent):
+    """
+    A class for managing time-dependent solvers.
+
+    This includes scenarios where the target is not directly a `TimeFunction`,
+    but depends on other functions that are.
+
+    Outline of time loop abstraction with PETSc:
+
+    - At PETScSolve, time indices are replaced with temporary `Symbol` objects
+      via a mapper (e.g., {t: tau0, t + dt: tau1}) to prevent the time loop
+      from being generated in the callback functions. These callbacks, needed
+      for each `SNESSolve` at every time step, don't require the time loop, but
+      may still need access to data from other time steps.
+    - All `Function` objects are passed through the initial lowering via the
+      `LinearSolveExpr` object, ensuring the correct time loop is generated
+      in the main kernel.
+    - Another mapper is created based on the modulo dimensions
+      generated by the `LinearSolveExpr` object in the main kernel
+      (e.g., {time: time, t: t0, t + 1: t1}).
+    - These two mappers are used to generate a final mapper `symb_to_moddim`
+      (e.g. {tau0: t0, tau1: t1}) which is used at the IET level to
+      replace the temporary `Symbol` objects in the callback functions with
+      the correct modulo dimensions.
+    - Modulo dimensions are updated in the matrix context struct at each time
+      step and can be accessed in the callback functions where needed.
+    """
+    @property
```

Comment:
Nitpick: should probably be `cached_property`, but tbh is so cheap it doesn't really matter